src/Week2/LogAnalysis.hs

```diff
@@ -0,0 +1,63 @@
+{-# OPTIONS_GHC -Wall #-}
+module LogAnalysis where
+
+import Text.Read
+import Week2.Log
+
+---------- ex 1 ----------
+
+parseError :: String -> LogMessage
+parseError string =
+        readError $ words string
+    where
+
+        helper :: MessageType -> [String] -> LogMessage
+        helper err (t:rest) =
+            case (readMaybe t :: Maybe Int) of
+                Nothing  -> Unknown string
+                Just int -> LogMessage err int $ unwords rest
+        helper _ _ = Unknown string
+
+        readError :: [String] -> LogMessage
+        readError ("I":xs) = helper Info xs
+        readError ("W":xs) = helper Warning xs
+        readError ("E":n:xs) =
+            case (readMaybe n :: Maybe Int) of
+                Nothing  -> Unknown string
+                Just int -> helper (Error int) xs
+        readError _ = Unknown string
+
+parse :: String -> [LogMessage]
+parse log = map parseError (lines log)
+
+---------- ex 2 ----------
+
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) tree = tree
+insert log Leaf = Node Leaf log Leaf
+insert log@(LogMessage _ time _) (Node left m@(LogMessage _ treeTime _ ) right)
+    | time < treeTime = Node (insert log left) m right
+    | time >= treeTime = Node left m (insert log right)
+
+---------- ex 3 ----------
+
+build :: [LogMessage] -> MessageTree
+build list = foldr (insert) Leaf list
+
+---------- ex 4 ----------
+
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf                      = []
+inOrder (Node left message right) = (inOrder left)++[message]++(inOrder right)
+
+---------- ex 5 ----------
+
+whatWentWrong :: [LogMessage] -> [String]
+whatWentWrong messages =
+        map getText sortedMessages
+    where
+        sortedMessages =  (dropWhile isBelow50) $ inOrder $ build messages
+        isBelow50 (LogMessage _ number _) = number < 50
+        getText (LogMessage _ _ text) = text
```

Comment:
Functions `isBelow50` and `getText` are not total. It would be better to rewrite code somehow to avoid non-exhaustive pattern-matching.

Usually it's a good idea to enable `-Wall` ghc option to see what's wrong with code.

Reply:
Yes I see the warnings for the incomplete pattern-matching, but I am not sure how to address it:

I am working with the type:
```
data LogMessage = LogMessage MessageType TimeStamp String
                | Unknown String
  deriving (Show, Eq)
```
How should I be handling cases where I know I will use the function `whatWentWrong` with a list `[LogMessage]` that I know will only have items of type `LogMessage MessageType TimeStamp String` (because of how the list was constructed)?

Should I still cover all cases? Wouldn't that be misleading as they are technically not supposed to happen? Should I cover the rest of the cases by throwing an error?

Reply:
In Haskell there's this popular concept: _Make invalid states unrepresentable_. So it's always a good idea to guarantee statically that something can't happen. How to do this in this particular case: another program design problem. But here is one way.

`dropWhile` function has type:

```haskell
dropWhile :: (a -> Bool) -> [a] -> [a]
```

It doesn't reflect in type that you don't have elements where given predicate returns `False`.

What you can use is `mapMaybe` function from `Data.Maybe` module:

* https://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Maybe.html#v:mapMaybe

It has the following type:

```haskell
mapMaybe :: (a -> Maybe b) -> [a] -> [b]
```

So if you pass some function with type `LogMessage -> Maybe Text` to `mapMaybe` you will get only those `Text`s where the given function returns `Just ...`. And inside this function you can return `Nothing` on `UnknownMessage` and if priority is below 50. Basically, you perform validation and transformation at the same time.